vpr/src/place/noc_place_utils.cpp

```diff
@@ -293,99 +391,159 @@ double comp_noc_aggregate_bandwidth_cost(void) {
+        temp_noc_routing_algorithm->route_flow(source_router_block_id, sink_router_block_id, temp_found_noc_route, noc_model);
 
         // now calculate the costs associated to the current traffic flow and accumulate it to find the total cost of the NoC placement
         double current_flow_aggregate_bandwidth_cost = calculate_traffic_flow_aggregate_bandwidth_cost(temp_found_noc_route, curr_traffic_flow);
-        noc_aggregate_bandwidth_cost_check += current_flow_aggregate_bandwidth_cost;
-
-        double current_flow_latency_cost = calculate_traffic_flow_latency_cost(temp_found_noc_route, *noc_model, curr_traffic_flow, noc_opts);
-        noc_latency_cost_check += current_flow_latency_cost;
+        cost_check.aggregate_bandwidth += current_flow_aggregate_bandwidth_cost;
+
+        auto [curr_traffic_flow_latency_cost, curr_traffic_flow_latency_overrun_cost] = calculate_traffic_flow_latency_cost(temp_found_noc_route, noc_model, curr_traffic_flow);
+        cost_check.latency += curr_traffic_flow_latency_cost;
+        cost_check.latency_overrun += curr_traffic_flow_latency_overrun_cost;
+
+        // increase bandwidth utilization for the links that constitute the current flow's route
+        for (auto& link_id : temp_found_noc_route) {
+            auto& link = temp_noc_link_storage[link_id];
+            double curr_link_bw_util = link.get_bandwidth_usage();
+            link.set_bandwidth_usage(curr_link_bw_util + curr_traffic_flow.traffic_flow_bandwidth);
+            VTR_ASSERT(link.get_bandwidth_usage() >= 0.0);
+        }
 
         // clear the current traffic flow route, so we can route the next traffic flow
         temp_found_noc_route.clear();
     }
 
+    // Iterate over all NoC links and accumulate congestion cost
+    for(const auto& link : temp_noc_link_storage) {
+        cost_check.congestion += calculate_link_congestion_cost(link);
+    }
+
     // check whether the aggregate bandwidth placement cost is within the error tolerance
-    if (fabs(noc_aggregate_bandwidth_cost_check - costs.noc_aggregate_bandwidth_cost) > costs.noc_aggregate_bandwidth_cost * error_tolerance) {
+    if (fabs(cost_check.aggregate_bandwidth - costs.noc_cost_terms.aggregate_bandwidth) > costs.noc_cost_terms.aggregate_bandwidth * error_tolerance) {
         VTR_LOG_ERROR(
             "noc_aggregate_bandwidth_cost_check: %g and noc_aggregate_bandwidth_cost: %g differ in check_noc_placement_costs.\n",
-            noc_aggregate_bandwidth_cost_check, costs.noc_aggregate_bandwidth_cost);
+            cost_check.aggregate_bandwidth, costs.noc_cost_terms.aggregate_bandwidth);
         error++;
     }
 
     // only check the recomputed cost if it is above our expected latency cost threshold of 1 pico-second, otherwise there is no point in checking it
-    if (noc_latency_cost_check > MIN_EXPECTED_NOC_LATENCY_COST) {
+    if (cost_check.latency > MIN_EXPECTED_NOC_LATENCY_COST) {
```

Comment:
Can we delete this first if?  latency_cost will always be above 0 (worth trying).

Reply:
I don't change the code in this branch because this condition has been removed in noc_turn_model_routing branch.